tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,242 @@
+        std::vector<int64_t> tb_unsharded_sizes       = {1       , num_devices_, 1                , params.K/num_devices_, params.N};
+        std::vector<int64_t> tb_sizes                 = {1       , 1           , 1                , params.K/num_devices_, params.N};
+        std::vector<int64_t> tc_locally_reduced_sizes = {params.S, 1           , params.M/params.S,                        params.N};
+        std::vector<int64_t> tc_sizes                 = {params.S, 1           , params.M/(params.S*num_devices_)        , params.N};
+    // clang-format on
+
+    // Set up input tensors. We create the full unsharded tensors and define the
+    // actual input as the shard corresponding to the current device. Having the
+    // full unsharded input on each rank makes it possible to compute the
+    // expected result locally, hence, this way of doing is convenient for
+    // validating data correctness.
+    at::TensorOptions options =
+        at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+    auto ta_unsharded = at::randn(ta_unsharded_sizes, options);
+    auto tb_unsharded = at::randn(tb_unsharded_sizes, options);
+    ta_ = at::empty(ta_sizes, options);
+    ta_.copy_(ta_unsharded.slice(1, my_device_index_, my_device_index_ + 1));
+    tb_ = at::empty(tb_sizes, options);
+    tb_.copy_(tb_unsharded.slice(1, my_device_index_, my_device_index_ + 1));
+
+    // We pre-allocate the output and some intermediate buffers so we do not
+    // rely on torch allocator, which do not behave well with multi-stream
+    // programming.
+    tc_locally_reduced_ = at::empty(tc_locally_reduced_sizes, options);
+    tc_ = at::empty(tc_sizes, options);
+
+    // compute the expected output for data correctness validation
+    auto tc_unsharded_unreduced = ta_unsharded * tb_unsharded;
+    auto tc_unsharded_expected = at::sum(tc_unsharded_unreduced, {1, 3});
+    auto tc_unsharded_expected_reshaped = at::reshape(
+        tc_unsharded_expected,
+        {params.S,
+         num_devices_,
+         params.M / (params.S * num_devices_),
+         params.N});
+    tc_expected_ = tc_unsharded_expected_reshaped.slice(
+        1, my_device_index_, my_device_index_ + 1);
+
+    // Debug print
+    if (communicator->deviceId() == 0 && debug_print) {
+      std::cout << params << std::endl;
+      std::cout << "ta_.sizes()=" << ta_.sizes() << std::endl;
+      std::cout << "tb_.sizes()=" << tb_.sizes() << std::endl;
+      std::cout << "tc_locally_reduced_.sizes()=" << tc_locally_reduced_.sizes()
+                << std::endl;
+      std::cout << "tc_.sizes()=" << tc_.sizes() << std::endl;
+    }
+  }
+
+  at::Tensor getSlice(at::Tensor t, int64_t j) {
+    return t.slice(0, j, j + 1);
+  }
+
+  void computeATen(
+      at::Tensor ta_j,
+      at::Tensor tb_j,
+      at::Tensor tc_locally_reduced_j) {
+    // we unsqueeze the output tensor to avoid a torch warning:
```

Comment:
Note to myself to try out later: I'm surprised this can't be avoided. Maybe tensordot is not as powerful as einsum.

Reply:
Ok. It works but gives a warning.
Lmk

Reply:
I pushed a change with an explanatory commit message. It's suboptimal either way so take it or leave it :)

Reply:
took it! thanks